### PR TITLE
added support for a new `Category` entity and missing functionalities

### PR DESCRIPTION
**Pull request by citruscaviar**

### CODE REVIEW

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+@RequestMapping("/categories")
+@RequiredArgsConstructor
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PostMapping
+    @Operation(description = "Creates a category for the book")
+    @PreAuthorize("hasRole('ADMIN')")
+    public CategoryResponseDto createCategory(@RequestBody @Valid CategoryRequestDto categoryDto) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:37]: ResponseStatus.CREATED

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        return categoryService.save(categoryDto);
+    }
+
+    @GetMapping("/{id}/books")
+    @Operation(description = "Finds books by category")
+    List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long id) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:43]:
```suggestion
    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long categoryId) {
```

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        return bookService.getBooksByCategory(id);
+    }
+
+    @GetMapping
+    @Operation(description = "Finds all categories")
+    public List<CategoryResponseDto> getAll(Pageable pageable) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:49]: better return Page instead List

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    @PostMapping
+    @Operation(description = "Creates a category for the book")
+    @PreAuthorize("hasRole('ADMIN')")
+    public CategoryResponseDto createCategory(@RequestBody @Valid CategoryRequestDto categoryDto) {
+        return categoryService.save(categoryDto);
+    }
+
+    @GetMapping("/{id}/books")
+    @Operation(description = "Finds books by category")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:42]: Let’s  add PreAuthorize (check other endpoints)

File: src/main/java/com/onlinebookstore/store/dto/CreateBookRequestDto.java [hunk 5]
```diff
@@ -24,4 +25,6 @@ public class CreateBookRequestDto {
     @Size
     private String description;
     private String coverImage;
+    @NotNull
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/dto/CreateBookRequestDto.java:28]: better use NotEmpty

File: src/main/java/com/onlinebookstore/store/repository/BookRepository.java [hunk 6]
```diff
@@ -1,7 +1,19 @@
...
 
 public interface BookRepository extends JpaRepository<Book, Long> {
+    @Query("FROM Book b JOIN b.categories c WHERE c.id = :categoryId")
+    List<Book> findAllByCategoryId(Long categoryId);
+
+    @Query("FROM Book b LEFT JOIN FETCH b.categories")
+    List<Book> findAllWithCategories(Pageable pageable);
+
+    @Query("FROM Book b LEFT JOIN FETCH b.categories WHERE b.id = :id")
+    Optional<Book> findBookById(Long id);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/repository/BookRepository.java:18]: Can you create a book without a category? 
Should you provide the option to create an uncategorized book?

File: src/main/java/com/onlinebookstore/store/service/CategoryService.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+    CategoryResponseDto getById(Long id);
+
+    CategoryResponseDto save(CategoryRequestDto categoryDto);
+
+    CategoryResponseDto update(Long id,
+                               CategoryRequestDto categoryDto);
+
+    void deleteById(Long id);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/service/CategoryService.java:19]:
```suggestion
```

File: src/main/java/com/onlinebookstore/store/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -26,22 +32,31 @@ public BookResponseDto save(CreateBookRequestDto requestDto) {
...
+        Book book = bookRepository.findBookById(id).orElseThrow(
                 () -> new EntityNotFoundException("Couldn't find a book with id: " + id));
         return bookMapper.toDto(book);
     }
 
     @Override
+    @Transactional
     public BookResponseDto update(Long id, CreateBookRequestDto requestDto) {
+        Set<Category> categories = categoryRepository
+                    .findByIdIn(requestDto.getCategoryIds());
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/service/impl/BookServiceImpl.java:56]: Better use getReferenceById (let’s google how it works)

```
private Set<Category> getCategoriesSet(Set<Long> categoriesIds) {
        return categoriesIds.stream()
                .map(categoryRepository::getReferenceById)
                .collect(Collectors.toSet());
    }
```

File: src/main/java/com/onlinebookstore/store/service/impl/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        Category categoryById = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Couldn't find a category by id: " + id)
+        );
+        categoryMapper.updateCategory(categoryDto,categoryById);
+        return categoryMapper.toDto(categoryRepository.save(categoryById));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/store/service/impl/CategoryServiceImpl.java:54]: Should you check if the category exists by ID?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.CREATED)
+    public CategoryResponseDto createCategory(@RequestBody @Valid CategoryRequestDto categoryDto) {
+        return categoryService.save(categoryDto);
+    }
+
+    @GetMapping("/{id}/books")
+    @Operation(description = "Finds books by category")
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long categoryId) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:46]:
```suggestion
    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long categoryId) {
```

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    @Operation(description = "Finds all categories")
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
+    public Page<CategoryResponseDto> getAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @GetMapping("/{id}")
+    @Operation(description = "Finds a category by id")
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
+    public CategoryResponseDto getCategoryById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:60]:
```suggestion
    public CategoryResponseDto getCategoryById(@PathVariable Long id) {
```

File: src/main/java/com/onlinebookstore/store/controller/CategoryController.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            @RequestBody @Valid CategoryRequestDto categoryDto
+    ) {
+        return categoryService.update(id, categoryDto);
+    }
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(description = "Deletes a category by id")
+    public void deleteCategory(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/store/controller/CategoryController.java:78]:
```suggestion
    public void deleteCategory(@PathVariable Long id) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**